File: solution/451. Sort Characters By Frequency/451_02.cpp
class Solution {
public:
    string frequencySort(string s) {
        int n = s.size();
        vector<int> freq(128);
        for (char& c : s)
            ++freq[c];
        
        vector<string> bucket(n + 1);
        for (int c = 0; c < 128; ++c)
            if (freq[c])
                bucket[freq[c]].push_back(c);
        
        string res;
        for (int i = n; i >= 1; --i)
            for (char& c : bucket[i])
                res.append(i, c);
        return res;
    }
};

File: solution/451. Sort Characters By Frequency/451_02.py
class Solution:
    def frequencySort(self, s: str) -> str:
        n = len(s)
        
        f = defaultdict(int)
        for c in s:
            f[c] += 1
        
        bucket = [[] for _ in range(n + 1)]
        for c, cnt in f.items():
            bucket[cnt].append(c)
        
        res = ""
        for i in range(n, 0, -1):
            for c in bucket[i]:
                res += i * c
        return res
    
